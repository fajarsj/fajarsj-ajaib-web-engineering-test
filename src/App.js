import { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";

import Layout from "./components/layout/Layout";
import Header from "./components/header/Header";
import Filter from "./components/filter/Filter";
import Table from "./components/table/Table";
import Pagination from "./components/pagination/Pagination";

import { getUserData } from "./store/user-actions";

let isInitial = true;

const App = () => {
  const dispatch = useDispatch();
  const users = useSelector((state) => state.user.data);
  const navigate = useNavigate();
  const [currentUser, setCurrentUser] = useState([]);
  const [selectedGender, setSelectedGender] = useState("");
  const [searchValue, setSearchValue] = useState("");

  useEffect(() => {
    dispatch(getUserData("?page=1&results=10"));
    navigate({
      search: "?page=1&results=10",
    });
  }, [dispatch, navigate]);

  useEffect(() => {
    if (isInitial) {
      isInitial = false;
      return;
    }

    setCurrentUser(users);
  }, [users, dispatch]);

  const filterByGender = (userArr, gender) => {
    return userArr.filter((user) => user.gender.toLowerCase() === gender);
  };

  const filterBySearch = (userArr, keyword) => {
    return userArr.filter((user) =>
      Object.keys(user).some((key) => user[key].toLowerCase().includes(keyword))
    );
  };

  useEffect(() => {
    let filterUsers = [...users];

    if (searchValue) {
      filterUsers = filterBySearch(filterUsers, searchValue);
    }

    if (selectedGender && selectedGender !== "all") {
      filterUsers = filterByGender(filterUsers, selectedGender);
    }

    setCurrentUser(filterUsers);
  }, [selectedGender, searchValue, users]);

  const handleSearchChange = (event) => {
    setSearchValue(event.target.value.toLowerCase());
  };

  const handleSelectChange = (event) => {
    setSelectedGender(event.target.value.toLowerCase());
  };

  const handleReset = () => {
    setSelectedGender("");
    setSearchValue("");
    setCurrentUser(users);
  };

  return (
    <Layout>
      <Header title="Example with search and filter" />
      <Filter
        handleSearchChange={handleSearchChange}
        handleSelectChange={handleSelectChange}
        handleReset={handleReset}
        selectedGender={selectedGender}
        searchValue={searchValue}
      />
      {users && <Table data={currentUser} />}
      <Pagination />
    </Layout>
  );
};

export default App;
