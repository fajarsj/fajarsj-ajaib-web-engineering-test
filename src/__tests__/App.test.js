import { renderWithProviders } from "../utils/test-utils";
import App from "../App";
import { BrowserRouter } from "react-router-dom";

test("render <App /> correctly", () => {
  const view = renderWithProviders(
    <BrowserRouter>
      <App />
    </BrowserRouter>
  );

  expect(view).toMatchSnapshot();
});
