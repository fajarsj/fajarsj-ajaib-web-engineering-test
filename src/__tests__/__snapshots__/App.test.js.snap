// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render <App /> correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="layout"
      >
        <div
          class="header"
        >
          <h1
            class="header__title"
          >
            Example with search and filter
          </h1>
        </div>
        <div
          class="filter"
        >
          <div
            class="input input--group"
          >
            <input
              class="input-control"
              placeholder="Search"
              type="text"
              value=""
            />
            <button
              class="btn btn--primary"
            >
              Search
            </button>
          </div>
          <select
            class="input-control"
          >
            <option
              disabled=""
              value=""
            >
              Select Gender
            </option>
            <option
              value="all"
            >
              All
            </option>
            <option
              value="male"
            >
              Male
            </option>
            <option
              value="female"
            >
              Female
            </option>
          </select>
          <button
            class="btn btn--outline-dark"
          >
            Reset Filter
          </button>
        </div>
        <table
          class="table"
        >
          <thead>
            <tr>
              <th>
                Username  
              </th>
              <th>
                Name  
              </th>
              <th>
                Email  
              </th>
              <th>
                Gender  
              </th>
              <th>
                Registered Date  
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                colspan="5"
              >
                <span
                  class="table__info"
                >
                  ⏳ Loading...
                </span>
              </td>
            </tr>
          </tbody>
        </table>
        <div
          class="pagination"
        >
          <button
            class="btn btn--outline-primary"
            disabled=""
          >
            Previous
          </button>
          <button
            class="btn btn--outline-primary"
            disabled=""
          >
            Next
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="layout"
    >
      <div
        class="header"
      >
        <h1
          class="header__title"
        >
          Example with search and filter
        </h1>
      </div>
      <div
        class="filter"
      >
        <div
          class="input input--group"
        >
          <input
            class="input-control"
            placeholder="Search"
            type="text"
            value=""
          />
          <button
            class="btn btn--primary"
          >
            Search
          </button>
        </div>
        <select
          class="input-control"
        >
          <option
            disabled=""
            value=""
          >
            Select Gender
          </option>
          <option
            value="all"
          >
            All
          </option>
          <option
            value="male"
          >
            Male
          </option>
          <option
            value="female"
          >
            Female
          </option>
        </select>
        <button
          class="btn btn--outline-dark"
        >
          Reset Filter
        </button>
      </div>
      <table
        class="table"
      >
        <thead>
          <tr>
            <th>
              Username  
            </th>
            <th>
              Name  
            </th>
            <th>
              Email  
            </th>
            <th>
              Gender  
            </th>
            <th>
              Registered Date  
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              colspan="5"
            >
              <span
                class="table__info"
              >
                ⏳ Loading...
              </span>
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="pagination"
      >
        <button
          class="btn btn--outline-primary"
          disabled=""
        >
          Previous
        </button>
        <button
          class="btn btn--outline-primary"
          disabled=""
        >
          Next
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`renders learn react link 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="layout"
      >
        <div
          class="header"
        >
          <h1
            class="header__title"
          >
            Example with search and filter
          </h1>
        </div>
        <div
          class="filter"
        >
          <div
            class="input input--group"
          >
            <input
              class="input-control"
              placeholder="Search"
              type="text"
              value=""
            />
            <button
              class="btn btn--primary"
            >
              Search
            </button>
          </div>
          <select
            class="input-control"
          >
            <option
              disabled=""
              value=""
            >
              Select Gender
            </option>
            <option
              value="all"
            >
              All
            </option>
            <option
              value="male"
            >
              Male
            </option>
            <option
              value="female"
            >
              Female
            </option>
          </select>
          <button
            class="btn btn--outline-dark"
          >
            Reset Filter
          </button>
        </div>
        <table
          class="table"
        >
          <thead>
            <tr>
              <th>
                Username  
              </th>
              <th>
                Name  
              </th>
              <th>
                Email  
              </th>
              <th>
                Gender  
              </th>
              <th>
                Registered Date  
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                colspan="5"
              >
                <span
                  class="table__info"
                >
                  ⏳ Loading...
                </span>
              </td>
            </tr>
          </tbody>
        </table>
        <div
          class="pagination"
        >
          <button
            class="btn btn--outline-primary"
            disabled=""
          >
            Previous
          </button>
          <button
            class="btn btn--outline-primary"
            disabled=""
          >
            Next
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="layout"
    >
      <div
        class="header"
      >
        <h1
          class="header__title"
        >
          Example with search and filter
        </h1>
      </div>
      <div
        class="filter"
      >
        <div
          class="input input--group"
        >
          <input
            class="input-control"
            placeholder="Search"
            type="text"
            value=""
          />
          <button
            class="btn btn--primary"
          >
            Search
          </button>
        </div>
        <select
          class="input-control"
        >
          <option
            disabled=""
            value=""
          >
            Select Gender
          </option>
          <option
            value="all"
          >
            All
          </option>
          <option
            value="male"
          >
            Male
          </option>
          <option
            value="female"
          >
            Female
          </option>
        </select>
        <button
          class="btn btn--outline-dark"
        >
          Reset Filter
        </button>
      </div>
      <table
        class="table"
      >
        <thead>
          <tr>
            <th>
              Username  
            </th>
            <th>
              Name  
            </th>
            <th>
              Email  
            </th>
            <th>
              Gender  
            </th>
            <th>
              Registered Date  
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              colspan="5"
            >
              <span
                class="table__info"
              >
                ⏳ Loading...
              </span>
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="pagination"
      >
        <button
          class="btn btn--outline-primary"
          disabled=""
        >
          Previous
        </button>
        <button
          class="btn btn--outline-primary"
          disabled=""
        >
          Next
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
